src/options.rs

```diff
@@ -1,228 +1,89 @@
+#[derive(TypedBuilder)]
+pub struct Exploration {
+  #[builder(
+    default = DEFAULT_POPULATION_SIZE, 
+    setter(doc = "Sets the population size. Increasing this option may produce higher quality neural
+                  networks, but will increase the time needed to find a solution."))]
+  pub population: usize,
+
+  #[builder(
+    default = DEFAULT_OFFSPRING_COUNT, 
+    setter(doc = "Sets the offspring count (how many offspring each individual spawns). Increasing this
+                  option may produce higher quality neural networks, but will increase the time needed to
+                  find a solution."))]
+  pub offspring: usize,
+
+  #[builder(
+    default = DEFAULT_SIMILAR_FITNESS, 
+    setter(doc= "Sets the threshold for deciding whether two neural networks have a similar fitness.
+                 Increasing this option will make the algorithm more aggresively prefer smaller
+                 neural networks. Decreasing it will do the opposite, allowing larger individuals to stay in
+                 the population. It is recommended to set this option higher if a small neural network is
+                 preferred. The downside is it will take slightly longer to find a solution, due to more
+                 higher fitness neural networks being discarded."))]
+  pub similar_fitness: f64,
+
+  #[builder(
+    default_code = "MutationSampler::default()", 
+    setter(doc = "Sets the sampler which chooses each mutation.  Build a `MutationSampler` with `MutationProbabilities`."))]
+  pub mutation_probabilities: MutationSampler,
+
+  #[builder(default = DEFAULT_EANT2_TERMINATION, setter(doc = "Termination conditions (target fitness, max generations)"))]
+  pub terminate: EANT2Termination,
+}
 
-    /// Sets the weight for choosing each mutation. The first element should be the weight for
-    /// adding a connection, the second for removing a connection, the third for adding a
-    /// neuron, and the fourth for adding a bias input. The weights are relative, so with the
-    /// weights `[1, 2, 4, 8]`, each mutation has twice the chance of the previous. In general,
-    /// the chance for connection mutations should be higher than the chance for a neuron mutation.
-    /// If a minimal network is desired, set the bias, neuron and connection addition weights low,
-    /// and the connection removal rate high. For complex problems where network size isn't an
-    /// issue, high connection addition rate is a good idea.
-    pub fn mutation_weights(mut self, weights: [usize; 4]) -> EANT2Options {
-        self.weights = weights;
-        self
-    }
+/// When should CMA-ES (inner loop) terminate?
+/// You usually don't need to configure this.
+#[derive(TypedBuilder)]
+pub struct CMAESTermination {
+  #[builder(default = None, setter(strip_option, doc = "Default is to use the fitness from `EANT2Termination` options, which defaults to `0.0`."))]
+  pub fitness:     Option<f64>,
```

Comment:
I don't think this option is necessary, as there doesn't seem to be any case where a user wouldn't want the default behavior here. Setting it to a looser requirement than `EANT2Termination.fitness` would make reaching the latter impossible, and setting it to a stricter requirement would cause CMA-ES to optimize more than required.

Also, could you add a `function_evals` option here that sets the `max_function_evals` option in `cmaes`? It would allow users to limit how many times the fitness function can be called per network optimization.

Reply:
Okay, makes sense.